src/Features/CSharp/Portable/CodeFixes/RemoveUnnecessaryCast/RemoveUnnecessaryCastCodeFixProvider.cs

```diff
@@ -21,89 +22,62 @@ namespace Microsoft.CodeAnalysis.CSharp.CodeFixes.RemoveUnnecessaryCast
 {
     [ExportCodeFixProvider(LanguageNames.CSharp, Name = PredefinedCodeFixProviderNames.RemoveUnnecessaryCast), Shared]
     [ExtensionOrder(After = PredefinedCodeFixProviderNames.ImplementInterface)]
-    internal partial class RemoveUnnecessaryCastCodeFixProvider : CodeFixProvider
+    internal partial class RemoveUnnecessaryCastCodeFixProvider : SyntaxEditorBasedCodeFixProvider
     {
-        public sealed override ImmutableArray<string> FixableDiagnosticIds
-        {
-            get { return ImmutableArray.Create(IDEDiagnosticIds.RemoveUnnecessaryCastDiagnosticId); }
-        }
-
-        public sealed override FixAllProvider GetFixAllProvider()
-        {
-            return RemoveUnnecessaryCastFixAllProvider.Instance;
-        }
-
-        private static CastExpressionSyntax GetCastNode(SyntaxNode root, SemanticModel model, TextSpan span, CancellationToken cancellationToken)
-        {
-            var token = root.FindToken(span.Start);
-            if (!token.Span.IntersectsWith(span))
-            {
-                return null;
-            }
-
-            return token.GetAncestors<CastExpressionSyntax>()
-                .FirstOrDefault(c => c.Span.IntersectsWith(span) && c.IsUnnecessaryCast(model, cancellationToken));
-        }
+        public sealed override ImmutableArray<string> FixableDiagnosticIds { get; } =
+            ImmutableArray.Create(IDEDiagnosticIds.RemoveUnnecessaryCastDiagnosticId);
 
         public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
         {
-            context.RegisterCodeFix(
-                new MyCodeAction(
-                    CSharpFeaturesResources.Remove_Unnecessary_Cast,
-                    c => RemoveUnnecessaryCastAsync(context.Document, context.Span, c)),
+            context.RegisterCodeFix(new MyCodeAction(
+                FeaturesResources.Remove_Unnecessary_Cast,
+                c => FixAsync(context.Document, context.Diagnostics.First(), c)),
                 context.Diagnostics);
             return SpecializedTasks.EmptyTask;
         }
 
-        private static async Task<Document> RemoveUnnecessaryCastAsync(
-            Document document, TextSpan span, CancellationToken cancellationToken)
+        protected override async Task FixAllAsync(
+            Document document, ImmutableArray<Diagnostic> diagnostics,
+            SyntaxEditor editor, CancellationToken cancellationToken)
         {
-            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
-            var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
-            var cast = GetCastNode(root, model, span, cancellationToken);
-            return await RemoveUnnecessaryCastAsync(document, cast, cancellationToken).ConfigureAwait(false);
+            var castNodes = diagnostics.SelectAsArray(
```

Comment:
so this does mix batch and serial fixing. nice.

Reply:
Yes.  In the IDE we should be using this for all our fixers.  I'm going through the process (including this PR) of trying to make that hte case.  "Simplify Type Name" is going to be the hardest.  But it will be one of the ones that would benefit the most here.